site/source/_posts/2017/11/17-Dojo-2-Beta-4/Dojo-2-Beta-4.md

```diff
@@ -0,0 +1,263 @@
+
+We have introduced a new accordion pane widget (`@dojo/widgets/accordionpane/AccordionPane`):
+
+<img src="./accordion.png" alt="The accordion pane widget" height="465" width="351" />
+
+As well as we have introduced a tooltip widget (`@dojo/widgets/tooltip/Tooltip`):
+
+<img src="./tooltip.png" alt="The tooltip widget" height="104" width="415" />
+
+## Other widget improvements
+
+We have also been trying to standardise our conventions in widgets in Beta 4, though we will likely have more changes for Beta 5.  In particular we have been rationalising our styles, making sure that themeable styles and the structural styles are clearly delineated, making it easier to theme existing widgets.  We have also been looking at the ways to better customise and extend the out of the box widgets.
+
+## Stores re-architecture
+
+The current API in Dojo 1 for managing state focused on a CRUD type API.  Originally in Dojo 2 we directly adapted that API as part of _stores_.  Recently though, the concept of state containers has taken front and centre and we knew we wanted to create a solution that aligned to that architecture.
+
+Instead of CRUD type API, it focuses on three main concepts:
+
+* _Operation_ which is a granular state mutation based on the JSON Patch API
+* _Command_ a semantically aligned function which returns a set of operations to apply
+* _Process_ a function that groups commands into a _process flow_ to represent a complete behaviour
+
+You can read more about the new API at [`dojo/stores`](https://github.com/dojo/stores).
+
+## A new approach to shims and polyfills
+
+We have rationalised the way we introduce compatibility across environments as well as improvements to the build system to optimize these when building an application.  We realised that some packages were wanting to depend upon 3rd party polyfills but we did not have a consistent way to introduce them as well as elide them if not needed.
+
+Now `@dojo/shim/main` loads all the polyfills needed to provide the expected base functionality and `@dojo/shim/browser` will load polyfills for a browser environment.  Most end developers will not need to worry about this as the other Dojo 2 packages will ensure that these modules get loaded.
+
+Of course if you are building an application that is targetting only certain platforms, these polyfills will simply result in a non-operation, but will still be part of the bundle.  We have added an additional static build optimisation feature, where a specially formed pragma will denote the feature that the next import will be polyfilling, and if that feature is not required based on the build targets, it will be removed from the build.  More information can be found in the [documentation for the feature](https://github.com/dojo/webpack-contrib#elided-imports).
```

Comment:
Of course if you are building an application that is targetting only certain platforms, these polyfills will simply result in a non-operation, but will still be part of the bundle.  We have added an additional static build optimisation feature, where a specially formed pragma will denote the feature that the next import will be polyfilling, and if that feature is not required based on the build targets, it will be removed from the build.  More information can be found in the [documentation for the feature](https://github.com/dojo/webpack-contrib#elided-imports). -> If you are building an application targeting only certain platforms, any unneeded polyfills will simply result in a non-operation, but will still be included in the bundle.  Beta 4 adds an additional static build optimization feature, where a specially formed pragma will denote the feature that the next import will be polyfilling. If that feature is not required based on the build targets, it will be removed from the build.  More information can be found in the [elided imports documentation](https://github.com/dojo/webpack-contrib#elided-imports).